### DISCUSSION TITLE
Gathering feedback about the Web API

### DISCUSSION THREAD

**Discussion by coreymckrill** [woocommerce/woocommerce#44199] [Request for Comments]
We’re looking to gather feedback from folks using the WooCommerce Web API to build tools, extensions, and more! 

As you may know, WooCommerce ships with several [server-side web APIs](https://en.wikipedia.org/wiki/Web_API#Server_side), collectively called the “Web API”. It includes the [REST API](https://woocommerce.github.io/woocommerce-rest-api-docs/), [Store API](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/src/StoreApi), webhooks, and AJAX endpoints.

If you build things that use the Web API, we’d love to hear your feedback. Specifically, we’d love to hear what, if anything, is currently missing and what could be improved. We’ll use this data to inform our product roadmap and to design future projects.

So, two questions, you can respond to one or both:

* What could be improved about the current Web API (REST endpoints, webhooks, AJAX)? For example, is there anything you find difficult or inefficient? Does anything not work the way you expect it to? Is there anything you would change?
* Is there anything you would add to the Web API? Why? Please include information about use-cases relating to this feature so we can best shape the roadmap.

You can also answer these questions in [survey form](https://woocommerce.survey.fm/web-api-survey).

**Comment by coreymckrill** [comment]
A few things that have come up so far in the survey responses:

* Faster response times and other performance improvements
* The ability to calculate shipping options before creating an order
* Different/more granular permission scopes
* Better documentation of error codes

**Comment by thejamescollins** [comment]
### TL;DR: We would like to see improvements with WooCommerce's webhook functionality and reliability.

## Context
We built and maintain WooCommerce Zapier, which sees (on average) 2.2+ million successful webhook deliveries across 36,000+ active webhooks each week.

At this scale, we encounter users with WooCommerce webhook reliability/deliverability issues on a daily basis. This causes confusion to WooCommerce users (data is not sent as expected), and it isn't something they can easily investigate or diagnose.

## Current State of WooCommerce Webhook Functionality

The 'happy path' for standard webhook operations is relatively well covered, however when errors or issues occur, there is not much visibility and there is no way of notifying the store owner.

For example:

- There is no (automatic or manual) retry mechanism for failed deliveries.
- [Webhook documentation](https://woo.com/document/webhooks/) isn't very detailed, and could be improved.
- Errors can occur without logs or notifications before or during the queuing of the webhook to the Action Scheduler.
- Delivery errors can occur when the Action Scheduler action is executed (eg timeouts, etc).
- Webhooks can send `WP_Error` payloads (for example #37806), and these are considered "successful" webhook deliveries.
- It is difficult for developers to investigate, diagnose and resolve webhook issues too.



## Suggestion: Full Webhook User Interface

I think the Stripe webhook mechanism and interface is a fantastic example of how webhook functionality can operate and be managed. Specifically:

- Each webhook listed, with error rates and a historical graph showing successful deliveries and tasks
- Each webhook delivery is stored in a database, and is full viewable, including the request body (payload) and the response code and body.
- Webhook deliveries can be filtered by status (Succeeded or Failed).
- Each webhook delivery can be resent or retried with the click of a button.


WooCommerce's webhook management screens should be improved to show error rates etc.

WooCommerce should introduce a new webhook activity database storage table so that each webhook delivery attempt and response is recorded. A dedicated database table for storing this is better than [a WooCommerce log file](https://woo.com/document/webhooks/#webhook-logs).

WooCommerce should display this delivery data in an easy to use interface inside the WooCommerce webhooks UI, and allow the user to retry failed deliveries manually if required.

## Suggestion: Built in (exponential back-off) retry mechanism

If a webhook delivery fails, it should be retried automatically using an exponential back-off mechanism similar to https://stripe.com/docs/webhooks#retries.


## Suggestion: Store Owner Notifications When Problems Occur

When a Stripe webhook repeatedly fails, an email is sent to the account owner alerting them that there is an issue. They can click through to view the webhook delivery history to identify and hopefully resolve the issue.

WooCommerce needs a way of notifying store owners when issues occur, so that they can be resolved in a timely fashion, instead of taking weeks or months before they realise there is a problem.

WooCommerce currently disables an active webhook if it encounters 5 consecutive errors. When this occurs, store owners aren't notified that there is an issue.

Store owners need to be told when these issues occur.

## Suggestion: Improve delivery reliability by queuing webhooks earlier

Currently, WooCommerce waits until the `shutdown` hook before adding pending webhook deliveries to the Action Scheduler queue (https://github.com/woocommerce/woocommerce/pull/27067).

This works well in many cases, but we have seen a growing number of cases where a site's `shutdown` hook does not fully execute and complete, resulting in data (silently) never being added to the webhook queue and thus never sent to Zapier.

Currently this is done at priority 10 (the default hook priority), which is also used by a *lot* of other plugins and extensions, who also do things at the end of the page load cycle.

The problem is that if any of those plugins does and `exit` or causes an error, the remaining `shutdown` tasks are never executed, resulting in WooCommerce's pending webhooks to never be added to the Action Scheduler queue.

We propose decreasing this to priority zero in order to increase the reliability of WooCommerce webhooks. This would likely be a simple change that should help avoid issues.

## Suggestion: Improve Webhook Documentation

The current [Webhook documentation](https://woo.com/document/webhooks/) should be expanded and improved upon, particuarly when it comes to troubleshooting webhook issues.

We have [documented many common webhook issues and solutions](https://docs.om4.io/woocommerce-zapier/troubleshooting/#data-not-sending-or-delayed) for our WooCommerce Zapier users, however most of the information applies to all WooCommerce webhook users (not just WooCommerce Zapier webhook users).

We would be happy for some (or all) of the above to be added to the WooCommerce official docs, so that all WooCommerce users can benefit.

# Thanks

Thanks to @om4csaba who helped me with the above thoughts and ideas.

# Questions?

We're keen to keen to help progress these ideas, and would love to hear from others who have any other ideas or comments on how WooCommerce's webhook mechanism and functionality can be improved.

**Reply by coreymckrill** [reply]
@om4james This is fantastic and thorough feedback, thank you!

> WooCommerce should introduce a new webhook activity database storage table so that each webhook delivery attempt and response is recorded. A dedicated database table for storing this is better than a WooCommerce log file.

I can definitely see the benefits of this! I'll just note some of the logistical things we'd likely need to consider if we were to implement something like this:

* How quickly would a high-volume site get into the millions of rows for this table? What would we do to mitigate table size if it became a problem?
* What kinds of PII might end up living in this table if we include "the request body (payload) and the response code and body"?
* As an alternative, would it still be useful if we just fired an action for each webhook delivery attempt and let plugins/developers who are interested in this data hook to that and pipe it into whatever system they want (e.g. Grafana dashboard or something)?

**Reply by bryceadams** [reply]
+1 to everything @om4james suggested. We at @metorikhq also process and work with millions of webhook deliveries on a daily basis and have experienced similar inconsistencies. Ultimately Woo cannot solve all of these, especially issues caused by 3rd party plugins, but more reliable webhook processing and a better interface for viewing webhook delivery stats would be a great start (Stripe's UI for this ❤️ ).

**Reply by thejamescollins** [reply]
hi @coreymckrill,

> How quickly would a high-volume site get into the millions of rows for this table? What would we do to mitigate table size if it became a problem?

Scheduled daily/weekly actions could be performed (via Action Scheduler) to clean up the results in this table over time.

For example, successful delivery data could be kept for 30 days (by default), and records older than that could either be removed entirely, or at least have their payload data removed to reduce storage size.

> What kinds of PII might end up living in this table if we include "the request body (payload) and the response code and body"?

This definitely depends on the resource, for example Orders would contain PII, but Coupons or Products would likely not.

Currently WooCommerce Webhook Logs [do not log the full webhook payload/body, unless `WP_DEBUG` is set to true](https://github.com/woocommerce/woocommerce/blob/e863c02551e94d928d3873131ff2f4ab61d0ff66/plugins/woocommerce/includes/class-wc-webhook.php#L567). So the same kind of approach could be used here too.

Importantly the Resource ID should always to be recorded, otherwise troubleshooting/debugging is very very difficult.

Unfortunately if original payloads aren't stored/saved, then that would limit the ability for one-click Retry/Resend options.

So a trade-off exists here, the convenience of more detailed debugging and automatic retry mechanisms vs the issue of storing unnecessary PII.

Another approach could be to store the full payload in the short term, and then once a webhook delivery has been successful (or after a pre-determined timeframe has elapsed), then remove the full payload.

This would have the short-term benefit of automatic retries, but avoid the longer term issues with unnecessary PII being stored long term.


> As an alternative, would it still be useful if we just fired an action for each webhook delivery attempt and let plugins/developers who are interested in this data hook to that and pipe it into whatever system they want (e.g. Grafana dashboard or something)?

The current webhook mechanism enables most of that now via these hooks:
- `woocommerce_webhook_delivery`
- `woocommerce_webhook_disabled_due_delivery_failures`

For completeness of this use-case, a new `woocommerce_webhook_delivery_failure` hook should be added to the `\WC_Webhook::failed_delivery()` method.



In our experience,  the default setup/configuration in WooCommerce matters greatly here. When WooCommerce Zapier customers contact us with a webhook issue, there is often very little audit trail stored already, making troubleshooting difficult.

So if WooCommerce could choose some sensible defaults such as keeping detailed webhook records for 30 days, with older records removed automatically, then that would help improve the experience for all users of webhooks.

**Reply by Hazmatyre** [reply]
I'd like to share the only repo I've found that has a GUI on firing webhooks off one wordpress hook at a time. It just boggles my mind how the built-in WooCommerce one is only exclusive to `woocommerce` hooks and I can't have something as simple as `save_post` for blog posts and custom pages.
https://github.com/fabiancdng/wp-hook-expose/

**Comment by coreymckrill** [comment]
A few more themes that have surfaced in the survey responses:

* GraphQL could provide efficiencies in some cases, such as retrieving several related objects with a specific subset of fields in one request
* Improved type safety in response data
* Better/more documentation in general (in addition to the previous mention of error code documentation)

**Comment by louwie17** [comment]
Thanks for starting this discussion @coreymckrill 
We have been making active use of the [REST API](https://woocommerce.github.io/woocommerce-rest-api-docs/#introduction) within the implementation of the new product editor. 

The REST API has served us quite well this far, we have made a few modifications/additions ourselves while working on the editor. Like adding a `suggested` products endpoint: https://github.com/woocommerce/woocommerce/pull/43720

And we have a couple bigger items we like to add ( there are probably more, but these are the first I could think of ):
- Add search support for name and SKU within the products rest endpoint -> https://github.com/woocommerce/woocommerce/issues/41662
- A REST endpoint to support the retrieval of meta data for custom fields -> https://github.com/woocommerce/woocommerce/issues/44169
- An endpoint that allows us to duplicate a product -> https://github.com/woocommerce/woocommerce/issues/43282 ( this one is quite important )

One big question we have had thus far in terms of contributing to the REST API is where do we draw the line for increasing the API version?
Currently we are going with the approach that if it doesn't break backwards compatibility we won't increase the API number. 
Is this the approach WooCommerce took for version 2 & 3? or is this documented anywhere?

**Reply by coreymckrill** [reply]
Thanks for chiming in @louwie17 !

> One big question we have had thus far in terms of contributing to the REST API is where do we draw the line for increasing the API version? Currently we are going with the approach that if it doesn't break backwards compatibility we won't increase the API number. Is this the approach WooCommerce took for version 2 & 3? or is this documented anywhere?

Unfortunately, I don't know the history behind the the current versions we have, and I haven't been able to unearth anything via the git commit history so far.

Versioning is definitely something we've been thinking about. Should we do a v4? Should we switch to some kind of date-based versioning, or base it on the version of WooCommerce itself? We don't have answers yet.

**Reply by joshuatf** [reply]
>One big question we have had thus far in terms of contributing to the REST API is where do we draw the line for increasing the API version?

WooCommerce has followed a pattern of reversioning endpoints when a breaking change is introduced.  This means with existing endpoints we should be able to:

* Add a new param to the returned response on any endpoints
* Add new optional params to the request of any endpoints
* Adding new endpoints

On the other hand, a breaking change includes:

* Removal of an endpoint
* Removal or a name change of a returned key in an endpoint
* Adding new required params to endpoint requests
* Changes to behavior in the endpoint (whether that's an underlying action, how data gets processed, or changes to the persisted data)

Personally, I'm strongly in favor of continuing to hold this strategy as it's worked well for us and allows enough flexibility for improvements in API while providing the strongest benefit of versioning: backwards compatibility.  I'm not aware of any issues we've had with this, but interested in hearing from others where problems were found using this style of versioning.

Also on a side note, this same logic does not apply to internal REST endpoints which are subject to breaking changes at any time.

**Reply by coreymckrill** [reply]
Not sure how big of a problem this really is, but one downside of making "non-breaking" changes while not updating the API version is that you can't be sure whether the REST server supports the params and endpoints you want to use, if they were added later. You have to either know the version of WooCommerce being run, or you have to make a separate request for the schema first.

**Reply by joshuatf** [reply]
@coreymckrill True, but the same issue exists even if the endpoint version is bumped.  You would still need to check if the latest endpoint version exists or which version of WC you're currently using before using that newly added endpoint.

**Comment by nathanss** [comment]
I've been using the `meta_data` property a lot while developing extensions for the new product editor. It's currently an array of objects ([reference](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-meta-data-properties)), which makes it a bit harder to create/update and delete entries. I think it would be easier if it was in the following format:

```js
{
  "meta_data": {
    [meta_key]: meta_value
  }
}
```

Or, if meta id is still needed:

```js
{
  "meta_data": {
    [meta_key]: {
      id: meta_id,
      value: meta_value
    }
  }
}
```

With this format, it would be easier to create, update, and delete values, as we wouldn't have to loop through an array.

**Reply by joshuatf** [reply]
I also feel this is a more expected use case for meta data in the endpoint.  I'm curious if any extensions are using the meta data with its current behavior to their benefit.

Also noting that in relation to [this comment](https://github.com/woocommerce/woocommerce/discussions/44199#discussioncomment-8409030), this falls under the `Changes to behavior in the endpoint` and would require a newly versioned endpoint to avoid breaks in backwards compatibility.

**Reply by senadir** [reply]
We also support meta data in Store API for Cart items but we make sure the API keys are consistent and stable, so we don't have a meta_key as a key. Why is doing an object of values easier than array of objects?

**Reply by joshuatf** [reply]
I'll let @nathanss speak to this, but I would speculate the answer is a matter of convenience in that it more closely aligns with how the data is stored in the client.  This just removes the need for mapping the key value pairs to an array of objects.

>but we make sure the API keys are consistent and stable

Is this referring to both the request and response being consistent or consistent with the data structure across other Woo endpoints?  Both of those make sense to me and seem like valid reasons to keep with the existing structure.

Out of curiosity, are there other benefits to the array of objects over the key that are used elsewhere or any background on why that structural choice was made?

**Reply by coreymckrill** [reply]
How would you handle a scenario where a meta key is not unique ("single" in WordPress parlance), i.e. there are multiple meta entries with the same key, but possibly different values?

**Reply by nathanss** [reply]
@senadir 

> Why is doing an object of values easier than array of objects?

With the current format, to find an existing entry, we have to loop through the array to find it. If it's an object, we can just check `meta_data[key]`. It's O(n), instead of O(1). Not a big deal if n is a small number, but I think it's something to keep in mind.

Updating the array with new entries is the trickiest, IMO. [Here's a code that does this](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-product-metadata.ts#L57). It's working well at the moment, but it took additional development time to make sure it was accounting for every edge case. 

@coreymckrill 

I haven't dealt with this type of metadata, so I haven't thought about it. Can you share an example of what type of data does that represent? Is that similar to an array? (multiple values assigned to a single key)

**Reply by coreymckrill** [reply]
@nathanss Off the top of my head, I don't have a real world example of multiple meta entries with the same key in a Woo object. But it's possible to do it, and I think that's why the `meta_data` property is structured the way it is, as an array of objects. The `meta_data` property is not specific to WooCommerce, it's how WordPress expresses meta data in the REST API for all sorts of different data objects.

You can try it out if you go to the screen to edit a product, open up the Screen Options tab at the top, and check the box to show Custom Fields. Then scroll down and add a couple of new custom field entries, both with the same key. Now if you retrieve the product data via `wc/v3/products`, you'll see both of the new meta entries in the `meta_data` field.

To @senadir's point, I think, if there's meta data that you need to get consistently, it's often best to "promote" it out of `meta_data` and into a top-level field in the response object. Since that's just adding a new field rather than changing one, that shouldn't break back-compat.

But! The reason I asked the question is that there very well could be other, better approaches too. We just need to make sure we accommodate how meta data actually works.

**Reply by nathanss** [reply]
@coreymckrill 

> To @senadir's point, I think, if there's meta data that you need to get consistently, it's often best to "promote" it out of meta_data and into a top-level field in the response object. Since that's just adding a new field rather than changing one, that shouldn't break back-compat.

I'll keep that in mind. I thought that the only reason to "promote" the field was to add some sort of API validation, but I guess there is a grey area in the middle. We can add to the API when it makes sense.

> But! The reason I asked the question is that there very well could be other, better approaches too. We just need to make sure we accommodate how meta data actually works.

If multiple meta entries with the same key are a thing, then I think it makes sense to keep the format as it is :)

**Reply by bryceadams** [reply]
We work with quite a lot of meta data through the REST API over at @metorikhq - the ID is quite important for tracking meta key/value pairs as there can be instances of duplicate keys etc. as @coreymckrill mentioned so I believe the API should include them. But when you store it on your side, you could opt to ignore the IDs and just store the keys/values if that data isn't needed.

**Comment by barryhughes** [comment]
> Better/more documentation in general (in addition to the previous mention of error code documentation)

Hi all! Interested if anyone has further thoughts on this (API documentation)?

- [This recent forum post](https://wordpress.org/support/topic/api-new-order-how-to-tell-woocommerce-not-to-send-back-all-order-data/) makes me think our overlap with the WordPress REST API may not be clear enough. Are there other gaps like this that we might not be aware of (particularly if you come from a non-WordPress background, you might be better placed to help us identify these)?
- Actual performance of [our existing docs site](https://woocommerce.github.io/woocommerce-rest-api-docs/) *isn't great* and I'm sure we can improve this. What else can we do to make this a better resource? I've seen some API docs that incorporate interactive sandboxes, for instance ... perhaps there are other features we could benefit from?

**Comment by gudh** [comment]
Hi, we have been using the REST API with nodejs package [@woocommerce/woocommerce-rest-api](https://www.npmjs.com/package/@woocommerce/woocommerce-rest-api), is there a plan to add TypeScript support to it?

**Reply by coreymckrill** [reply]
@gudh Just to make sure I understand the question, are you wondering if TypeScript type definitions could be added to the package?

**Reply by gudh** [reply]
@coreymckrill 

Yeah, but doing so probably requires rewriting the package

There's existing type package like [@types/woocommerce__woocommerce-rest-api](https://www.npmjs.com/package/@types/woocommerce__woocommerce-rest-api), however, because of the way existing package is written, all the api calls return `any` type, which is basically meaningless, simply avoids TypeScript error complain

Ideally, when you call a woo product endpoint, you return `WooProduct` type, and when you call a woo order endpoint, you return `WooOrder` type, things like that

Thanks for starting this discussion, I am looking forward to the new Web API!

**Comment by mdperez86** [comment]
Hey, in Mothra we are looking for https://github.com/woocommerce/woocommerce/issues/41662 to be implemented as a product team requirement. We need the [list product endpoint](https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products) to let us search products by `name` AND/OR `sku`, currently it is not possible, we can search by `sku` but not by many product properties including the `sku`. After taking a look at the implementation we found that the feature was intentionally removed/avoided. We need this feature. Let us know what are the next steps for this so we can help on it.
cc: @pmcpinto @louwie17

**Reply by coreymckrill** [reply]
Hi @mdperez86, you may have already found this, but https://github.com/woocommerce/woocommerce/pull/32046 is a previous PR related to product search that gives a bit of background. One thing that's only sort of mentioned there is that searching by multiple fields at once (like name and sku) has performance implications because you have to have extra join clauses in the database query. So we need to tread carefully in implementing this. Probably best to discuss solutions over in #41662.

More generally, I'm seeing your feedback here as "Searching for specific products/orders/customers/etc via the REST API is inefficient and/or difficult." Would you say that's accurate?

**Comment by mdperez86** [comment]
Hey, when working on scheduling a product we realized that `date_created` property is used to determine the schedule state of the product (status = 'future') and we identify some problems there.
1. When the site timezone is different from `UTC-0` if we send 
    ```javascript
    // POST: /wp-json/wc/v3/products/4732
    {
        "id": 4732,
        "date_created": "2024-02-21T13:05:21"
    }
    ``` 
    we receive this back 
    ```javascript
    {
        "id": 4732,
        "date_created": "2024-02-21T10:05:21",
        "date_created_gmt": "2024-02-21T13:05:21",
        "date_modified": "2024-02-20T13:05:37",
        "date_modified_gmt": "2024-02-20T16:05:37",
        "status": "future",
        ...
    }
    ```
    as you can see the `date_created` differs between the request and the response and it should not. For some reason the server applies the timezone conversion, this datetime should not be touched. This is not happening if we send the specific timezone as part of the date string. And this is not the behavior when creating/editing a post.

3. Using a `date_created` for scheduling can be confusing, maybe it's a naming problem here, but the moment a product is created should be different form the moment a product is scheduled. We know the Post uses `post_date` property for the same purpose but the name in this case is not specific like in the Product. `date_created` can be an interesting metric for Analytics as well as `date_scheduled` as separate metrics and that is another reason why we think we should not use `date_created` for scheduling purposes.

Let us know what are the next steps for this so we can help on it.

**Reply by coreymckrill** [reply]
Hi @mdperez86, for the first point, I may not be understanding correctly, but I think this is the expected behavior when not specifying the timezone. Regardless, please open a GitHub issue about this (with `focus: rest api` label) and we can figure it out there.

Regarding the second point, this seems like it's deeper than the Web API, and is a more fundamental limitation of the way that the products data structure is currently implemented. But it's still good feedback, and something we should consider for any future improvements to products.

**Comment by erikdemarco** [comment]
for the webhook there should be "triggered_by" data. So we can know which triggered this event.

This is to prevent infinite loop in two way sync when using automation such as zapier.

For example:
Zap 1: When an order edited in woocommerce, edit the same order in Shopify.
Zap 2: When an order edited ib Shopify, edit the same order in Woocommerce.

This will run forever. But if we know this 'order updated' event was triggered by zapier, we can stop processing it.

**Reply by thejamescollins** [reply]
When a WooCommerce webhook is delivered, it also contains some HTTP headers with this information in it. For example:

- `X-WC-Webhook-Topic` => `order.updated`
- `X-WC-Webhook-Resource` => `order`
- `X-WC-Webhook-Event` => `updated`

So I think the information you're looking for is in the webhook delivery headers (and not in the webhook payload).

Does that suit your needs?

**Reply by erikdemarco** [reply]
@om4james And how that information can help preventing the infinite loop in 2 way sync scenario?

**Comment by vedanshujain** [comment]
# Suggestion: ChangeLog API
 This is quite similar to suggestion about WebHooks, just from a different interface point of view.
 
## Use case

Many users of REST APIs and webhooks, such as mobile Apps, analytical or marketing systems, JP sync etc, are interested in finding out what changed since their last sync. These systems would maintain a copy of relevant data locally and will try to update them by querying REST API periodically, or listen to WebHooks. 

While WebHook presumably is a sufficient system for this, it has some drawbacks:

1. It needs an always available domain to receive those WebHooks. As such it's not a great fit for systems such as mobile Apps.
2. It can be unreliable, for example, when either the site sending the webhook, or the site that is receiving it, is down.
3. It's not very efficient, as WooCommerce might send webhook individually for every change, even under high load. This blocks PHP workers while they are just waiting on web request to be completed and can adversaly effect the performance of the site.

REST API in it's current form is also not a great fit, because we currently don't provide good ways to just get the changes after last sync.

## Possible solution

We should consider introducing a changelog API with following features:

1. Get all the changes for a resource since they last synced. This should be available for all the resources that are available via REST API.

2. Merchants should be able to see the apps that registered for querying latest changes, and be able to revoke access to them. This is important for them to know where their data is being cached locally.

3. It should be efficient. Just one request should be able to get paginated details of many records that are changed, as opposed to webhook which sends individual requests for every change.

**Comment by 3lev8** [comment]
I'm currently experiencing the following:

My supplier provides me with category id's for parent / child relation and to assign products to categories.
However, the ID is read only in REST API so I'm looking for the best way to make this work.

Would be cool If there would be a custom ID field for these kinds of situations.
I understand there could be a conflict with existing post id's but this could be easily fixed by for example multiplying the ID by 10.

**Reply by coreymckrill** [reply]
@3lev8 I'm not sure I understand the enhancement you'd like to see here. Are you talking about custom category IDs or custom product IDs? Or something else?

**Comment by the-csaba** [comment]
Hello everyone,

Has there been an intended timeframe for collecting feedback? Perhaps more crucially, is there any opportunity to shape the project based on this feedback?

It would be incredibly helpful to learn when the feedback collection and outcomes will conclude. In my responses to the feedback form, I asked about similar aspects (scope, timeframe, effort, etc.), but I am still awaiting updates. While I'm not expecting individual follow-ups, some general guidance would be greatly appreciated.

This is particularly important for us at OM4 because our Zapier integration relies heavily on the REST API. Any changes could significantly influence our development roadmap.

Is there a more suitable forum for discussing these concerns, or is this the correct place? Recent discussions have primarily focused on troubleshooting, which, to my understanding, was not the original purpose.

We would be open to collaboration or a direct conversation with @coreymckrill or any other member of the WooCommerce team.

Thanks

**Reply by coreymckrill** [reply]
@om4csaba your timing in asking this question is impeccable. We just published a summary of our findings from the survey and related discussions [here](https://developer.woo.com/2024/03/14/exploring-the-future-of-the-woocommerce-web-api/). To answer your questions more specifically, as well as I can right now:

> Has there been an intended timeframe for collecting feedback?

We were roughly aiming to collect feedback through the month of February. It ended up going over a bit into March, but we've closed the survey now, and we haven't gotten any new responses there for a while. Eventually we might close this discussion too (in favor of future ones about more specific Web API topics) but seems fine to leave it open for now.

> Perhaps more crucially, is there any opportunity to shape the project based on this feedback?

I think there will be, probably through more GitHub discussions and perhaps some other avenues. We're still figuring out what that might look like.

> This is particularly important for us at OM4 because our Zapier integration relies heavily on the REST API. Any changes could significantly influence our development roadmap.

We would prefer that any changes we end up making to the Web API don't come as a big surprise to anyone, so everyone has time to try things out and give feedback as we go.

> Is there a more suitable forum for discussing these concerns, or is this the correct place?

I think here in GitHub discussions is a good place. You can open your own discussion topics here if there are specific things you'd like to talk through.

**Reply by the-csaba** [reply]
Thank you for the comprehensive update, @coreymckrill!

I appreciate your timely response and the detailed summary provided in the link. It's reassuring to know the intended timeframe for collecting feedback and the potential for shaping the project based on this feedback. 

I'm looking forward to further discussions on this platform and exploring how we can contribute to and prepare for any upcoming changes to the Web API. 

Thanks again for keeping the lines of communication open and transparent, and I look forward to more fruitful discussions soon.
